design/service-discovery.md

```diff
@@ -0,0 +1,178 @@
+
+As with the server side, when a new request subscriber is discovered, the GUID of the corresponding response publisher is taken from the `USER_DATA` and the mapping between the request subscriber and response publisher is stored.
+Also, when creating the request publisher, the response subscriber's GUID is added to the `USER_DATA` for the server to be able to make the connection between them.
+
+### rmw_create_client ###
+
+This method already creates the response subscriber first and the request publisher afterwards. The only required change is to get the subscriber's GUID and store it on the `USER_DATA` on the `publisherParam` before creating the publisher.
+
+### CustomParticipantInfo ###
+
+* Add `remote_response_guid_by_remote_request_guid_` to hold the relation between a request subscriber and its corresponding response publisher.
+  It will be an unordered map with the request subscriber GUID as key:
+  `std::unordered_map<eprosima::fastrtps::rtps::GUID_t, eprosima::fastrtps::rtps::GUID_t, rmw_fastrtps_shared_cpp::hash_fastrtps_guid>`
+
+  **Note:** This is the same object as before, it will be reused in both the client and server sides.
+
+### ParticipantListener::onSubscriberDiscovery ###
+
+* When a new subscriber is discovered, read the `USER_DATA` to get the corresponding response publisher's GUID and store the relation on the `remote_response_guid_by_remote_request_guid_` map.
+* If the `USER_DATA` does not contain a GUID, **do not add an entry to the map**. This signals that the remote endpoint is not compiant with these modifications (for backwards compatibility and with other vendors).
+* When a subscriber is un-discovered, remove the entry from the `remote_response_guid_by_remote_request_guid_` map (if any).
+
+![ParticipantListener::onSubscriberDiscovery](service-discovery-images/ParticipantListener-onSubscriberDiscovery.svg)
+
+### CustomClientInfo ###
+
+* Add a reference to CustomParticipantInfo, to be able to retrieve the relations between the incoming request and the subscriber that will be receiving the response.
+
+* Add `pending_matches_` to keep track of cases where the request publisher has matched with the server but the response subscriber has not. It will be an unordered map with the server response publisher's GUID as key:
+  `std::unordered_map<eprosima::fastrtps::rtps::GUID_t, eprosima::fastrtps::rtps::GUID_t, rmw_fastrtps_shared_cpp::hash_fastrtps_guid>`
+
+* Add `complete_matches_` to keep track of fully matched servers. It will be an unordered map with the server response publisher's GUID as key:
+  `std::unordered_map<eprosima::fastrtps::rtps::GUID_t, eprosima::fastrtps::rtps::GUID_t, rmw_fastrtps_shared_cpp::hash_fastrtps_guid>`
+
+* Add `complete_matches_count_` to hold the size of `complete_matches_`, to be used on `rmw_service_server_is_available`. It will be an atomic variable `std::atomic_size_t`.
```

Comment:
@IkerLuengo this is meant for locks during `complete_matches_` updates not to penalize `rmw_service_server_is_available()` calls, right? If so, a one sentence explanation would be nice.